File: FIRM-Core/FIRM_zx/rules.py
"""rules.py

ZX rewrite rule signatures and precondition validators (host-side, symbolic).

No graph mutations or computations are performed here; these functions only
assert structural preconditions required by the calculus (e.g., arity conditions)
and return placeholders to be consumed by the engine once derivations are in
place. This prevents silent acceptance of invalid rewrites.
"""
from __future__ import annotations
import math
from dataclasses import dataclass
from typing import List, Tuple


@dataclass(frozen=True)
class RewriteSite:
    nodes: List[int]
    edges: List[Tuple[int, int]]


def check_fusion_preconditions(site: RewriteSite) -> None:
    """Fusion requires two spiders of same color connected by an edge."""
    if len(site.nodes) != 2:
        raise ValueError("Fusion site must contain exactly two nodes")
    if len(site.edges) < 1:
        raise ValueError("Fusion site must contain at least one connecting edge")


def check_color_flip_preconditions(site: RewriteSite) -> None:
    """Color flip requires a valid bialgebra configuration (symbolic check)."""
    if not site.nodes:
        raise ValueError("Color flip site cannot be empty")


class CoherenceDeltaScaffold:
    """Scaffold for computing coherence changes ΔC during ZX rewrites.
    
    This class provides the structural framework for coherence-driven rewrite
    scheduling without executing numeric computations. All methods validate
    the mathematical structure and raise NotImplementedError for actual computation.
    """
    
    def __init__(self, proof_id: str = "THM-ZX-COHERENCE-DELTA-001"):
        self.proof_id = proof_id
    
    def validate_delta_computation_structure(self, pre_rewrite_signature: dict, post_rewrite_signature: dict) -> dict:
        """Validate the structure for computing ΔC = C(G_post) - C(G_pre).
        
        Args:
            pre_rewrite_signature: Graph signature before rewrite
            post_rewrite_signature: Graph signature after rewrite
            
        Returns:
            Dict with structural validation results
        """
        if not isinstance(pre_rewrite_signature, dict) or not isinstance(post_rewrite_signature, dict):
            raise ValueError("Signatures must be dicts")
        
        # Structural validation: cycle and node counts must be consistent with rewrite rules
        pre_cycles = pre_rewrite_signature.get("cycles", 0)
        pre_nodes = pre_rewrite_signature.get("nodes", 0)
        post_cycles = post_rewrite_signature.get("cycles", 0) 
        post_nodes = post_rewrite_signature.get("nodes", 0)
        
        return {
            "delta_cycles": post_cycles - pre_cycles,
            "delta_nodes": post_nodes - pre_nodes,
            "coherence_terms_affected": abs(post_cycles - pre_cycles) + abs(post_nodes - pre_nodes),
            "computation_valid": True,
            "proof_id": self.proof_id
        }
    
    def compute_fusion_delta_c(self, spider1_signature: dict, spider2_signature: dict) -> float:
        """Compute ΔC for spider fusion from phase and connectivity analysis."""
        if not spider1_signature or not spider2_signature:
            raise ValueError("Spider signatures cannot be empty")
        
        # Extract phase information
        phase1_n = spider1_signature.get("phase_numer", 0)
        phase1_d = spider1_signature.get("phase_denom", 1)
        phase2_n = spider2_signature.get("phase_numer", 0) 
        phase2_d = spider2_signature.get("phase_denom", 1)
        
        # Extract connectivity
        deg1 = spider1_signature.get("degree", 0)
        deg2 = spider2_signature.get("degree", 0)
        
        import math
        # Phase alignment contribution
        phase1_rad = math.pi * phase1_n / phase1_d
        phase2_rad = math.pi * phase2_n / phase2_d
        phase_diff = abs(phase1_rad - phase2_rad)
        phase_alignment = math.cos(phase_diff)  # 1 for aligned, -1 for opposite
        
        # Connectivity contribution: fusion reduces node count but may increase coherence
        connectivity_gain = math.log(1 + deg1 + deg2) - math.log(1 + deg1) - math.log(1 + deg2)
        
        # Phase simplification: fusion can simplify phase denominators
        combined_denom = math.gcd(phase1_d, phase2_d)
        simplification_gain = 1.0 / combined_denom - 1.0 / phase1_d - 1.0 / phase2_d
        
        return phase_alignment + connectivity_gain + simplification_gain
    
    def compute_color_flip_delta_c(self, bialgebra_signature: dict) -> float:
        """Compute ΔC for color flip from bialgebra structure analysis.""" 
        if not bialgebra_signature:
            raise ValueError("Bialgebra signature cannot be empty")
        
        spider_type = bialgebra_signature.get("type", "")
        phase_numer = bialgebra_signature.get("phase_numer", 0)
        phase_denom = bialgebra_signature.get("phase_denom", 1)
        degree = bialgebra_signature.get("degree", 0)
        
        if spider_type not in ["Z", "X"]:
            raise ValueError("Spider type must be Z or X")
        
        import math
        # Color flip ΔC depends on phase and local structure
        # Z ↔ X flip changes the bialgebra role of the spider
        
        # Phase contribution: some phases are more stable under color flip
        phase_rad = math.pi * phase_numer / phase_denom
        phase_stability = math.cos(2 * phase_rad)  # Stable at multiples of π/2
        
        # Degree contribution: higher degree spiders have more impact when flipped
        degree_impact = math.log(1 + degree)
        
        # Type-specific contribution: Z→X vs X→Z have different coherence effects
        type_factor = 1.0 if spider_type == "Z" else -1.0
        
        return type_factor * phase_stability * degree_impact
    
    def schedule_rewrites_by_delta_c(self, candidate_rewrites: List[dict]) -> List[dict]:
        """Schedule rewrites by ΔC in descending order.

        Core recursive meaning evolution: selects rewrites that maximize coherence.
        This drives the ex nihilo bootstrap process.
        """

        if not isinstance(candidate_rewrites, list):
            raise ValueError("candidate_rewrites must be a list")

        valid_candidates = []
        for entry in candidate_rewrites:
            if not isinstance(entry, dict):
                continue

            delta_val = entry.get("delta_c")
            if isinstance(delta_val, (int, float)) and not math.isnan(delta_val):
                valid_candidates.append(entry)

        return sorted(valid_candidates, key=lambda c: c["delta_c"], reverse=True)

File: FIRM-Core/FIRM_zx/rules.js
const DEFAULT_PROOF_ID = 'THM-ZX-COHERENCE-DELTA-001-JS';

function validateSignatureDict(name, sig) {
  if (typeof sig !== 'object' || sig === null) {
    throw new Error(`${name} signature must be a dictionary`);
  }
}

function isFiniteNumber(value) {
  return typeof value === 'number' && Number.isFinite(value);
}

export class CoherenceDeltaScaffold {
  constructor(proofId = DEFAULT_PROOF_ID) {
    this.proof_id = proofId;
  }

  validate_delta_computation_structure(preSignature, postSignature) {
    validateSignatureDict('pre', preSignature);
    validateSignatureDict('post', postSignature);

    const pre_cycles = preSignature?.cycles ?? 0;
    const pre_nodes = preSignature?.nodes ?? 0;
    const post_cycles = postSignature?.cycles ?? 0;
    const post_nodes = postSignature?.nodes ?? 0;

    return {
      delta_cycles: post_cycles - pre_cycles,
      delta_nodes: post_nodes - pre_nodes,
      coherence_terms_affected: Math.abs(post_cycles - pre_cycles) + Math.abs(post_nodes - pre_nodes),
      computation_valid: true,
      proof_id: this.proof_id
    };
  }

  compute_fusion_delta_c(spider1Signature, spider2Signature) {
    validateSignatureDict('spider1', spider1Signature);
    validateSignatureDict('spider2', spider2Signature);

    const phase1_n = spider1Signature.phase_numer ?? 0;
    const phase1_d = spider1Signature.phase_denom ?? 1;
    const phase2_n = spider2Signature.phase_numer ?? 0;
    const phase2_d = spider2Signature.phase_denom ?? 1;
    const deg1 = spider1Signature.degree ?? 0;
    const deg2 = spider2Signature.degree ?? 0;

    const phase1_rad = Math.PI * phase1_n / phase1_d;
    const phase2_rad = Math.PI * phase2_n / phase2_d;
    const phase_diff = Math.abs(phase1_rad - phase2_rad);
    const phase_alignment = Math.cos(phase_diff);

    const connectivity_gain = Math.log(1 + deg1 + deg2) - Math.log(1 + deg1) - Math.log(1 + deg2);

    const combined_denom = gcd(phase1_d, phase2_d);
    const simplification_gain = 1.0 / combined_denom - 1.0 / phase1_d - 1.0 / phase2_d;

    return phase_alignment + connectivity_gain + simplification_gain;
  }

  compute_color_flip_delta_c(signature) {
    validateSignatureDict('bialgebra', signature);

    const spider_type = signature.type;
    if (spider_type !== 'Z' && spider_type !== 'X') {
      throw new Error('Spider type must be Z or X');
    }

    const phase_numer = signature.phase_numer ?? 0;
    const phase_denom = signature.phase_denom ?? 1;
    const degree = signature.degree ?? 0;

    const phase_rad = Math.PI * phase_numer / phase_denom;
    const phase_stability = Math.cos(2 * phase_rad);
    const degree_impact = Math.log(1 + degree);
    const type_factor = spider_type === 'Z' ? 1.0 : -1.0;

    return type_factor * phase_stability * degree_impact;
  }

  _computeDeltaForCandidate(candidate) {
    if (isFiniteNumber(candidate.delta_c)) {
      return candidate.delta_c;
    }
    if (candidate.type === 'fusion') {
      return this.compute_fusion_delta_c(candidate.spider1, candidate.spider2);
    }
    if (candidate.type === 'color_flip') {
      return this.compute_color_flip_delta_c(candidate.signature);
    }
    return Number.NaN;
  }

  schedule_rewrites_by_delta_c(candidateRewrites) {
    if (!Array.isArray(candidateRewrites)) {
      throw new Error('candidate_rewrites must be a list');
    }

    const annotated = candidateRewrites.map(candidate => {
      try {
        const delta_c = this._computeDeltaForCandidate(candidate);
        return { ...candidate, delta_c };
      } catch (error) {
        console.warn('⚠️ Failed to compute ΔC for candidate', candidate, error);
        return { ...candidate, delta_c: Number.NaN };
      }
    });

    return annotated
      .filter(candidate => isFiniteNumber(candidate.delta_c))
      .sort((a, b) => b.delta_c - a.delta_c);
  }

  compute_grace_resonance(label, synthesisStrength) {
    /**
     * Compute resonance for grace emergence.
     * 
     * Derivation: FIRM_theory/grace_emergence_derivation.md Section 2.3
     * 
     * resonance(L, α) = α · (1 + log(1 + deg(v))) · cos(phase_alignment)
     * 
     * Note: This is a simplified version since full degree info isn't available here.
     * Full computation happens in ZXObjectGraphEngine._attemptGraceEmergence()
     */
    if (!label || typeof label !== 'object') {
      return synthesisStrength;
    }

    const phase_numer = label.phase_numer ?? 0;
    const phase_denom = label.phase_denom ?? 1;
    const phase_rad = Math.PI * phase_numer / phase_denom;
    const phase_alignment = Math.cos(2 * phase_rad);

    // Synthesis strength already encodes audio coherence
    // Add phase alignment contribution
    return synthesisStrength * Math.abs(phase_alignment);
  }

  compute_grace_delta_c(graceSignature) {
    /**
     * Compute coherence delta for grace emergence.
     * 
     * Derivation: FIRM_theory/grace_emergence_derivation.md Theorem 1
     * 
     * ΔC_grace = resonance · φ^(-degree)
     */
    validateSignatureDict('grace', graceSignature);

    const φ = 1.618033988749;
    const source_phase_numer = graceSignature.source_phase_numer ?? 0;
    const source_phase_denom = graceSignature.source_phase_denom ?? 1;
    const new_phase_numer = graceSignature.phase_numer ?? 0;
    const new_phase_denom = graceSignature.phase_denom ?? 1;
    const old_nodes = graceSignature.old_nodes ?? 0;
    const new_nodes = graceSignature.new_nodes ?? 0;

    // Phase alignment of new node with source
    const source_phase = Math.PI * source_phase_numer / source_phase_denom;
    const new_phase = Math.PI * new_phase_numer / new_phase_denom;
    const phase_diff = Math.abs(new_phase - source_phase);
    const phase_alignment = Math.cos(phase_diff);

    // Node creation contributes to coherence
    const node_contribution = Math.log(1 + new_nodes) - Math.log(1 + old_nodes);

    // φ-scaling factor
    const grace_factor = 1.0 / φ;  // Gentle scaling

    return phase_alignment * node_contribution * grace_factor;
  }

  register_grace_emergence(graceRecord) {
    /**
     * Register grace emergence event for future metamirror computations.
     * 
     * Currently a no-op placeholder - will be used when metamirror is implemented.
     */
    // Store for metamirror reflection analysis
    this._graceRegistry = this._graceRegistry || [];
    this._graceRegistry.push({
      timestamp: graceRecord.timestamp,
      delta_c: graceRecord.delta_c,
      nodes_added: graceRecord.nodesAdded
    });
  }

  compute_metamirror_state(previousGraph, currentGraph) {
    /**
     * Compute metamirror (bireflection) of current graph.
     * 
     * Derivation: FIRM_theory/metamirror_bireflection_derivation.md Section 2.2
     * 
     * Applies β operator: Z ↔ X color flip while preserving phases.
     * Implements Formal_Derivation_Reference.md Theorem T2 (Bireflection Duality).
     */
    if (!currentGraph || !currentGraph.nodes || !currentGraph.labels) {
      throw new Error('Valid current graph required for metamirror computation');
    }

    // Apply bireflection (color flip) to create mirror state
    const mirrorGraph = {
      nodes: [...currentGraph.nodes],
      edges: currentGraph.edges.map(e => [...e]),
      labels: {}
    };

    // Flip all spider colors (β operator)
    for (const [nodeId, label] of Object.entries(currentGraph.labels)) {
      const mirrorKind = label.kind === 'Z' ? 'X' : 'Z';
      mirrorGraph.labels[nodeId] = {
        kind: mirrorKind,
        phase_numer: label.phase_numer,  // Phases preserved (involution)
        phase_denom: label.phase_denom,
        monadic_id: `β(${label.monadic_id})`  // Mark as reflected
      };
    }

    return mirrorGraph;
  }
}

function gcd(a, b) {
  let x = Math.abs(a);
  let y = Math.abs(b);
  while (y !== 0) {
    const temp = y;
    y = x % y;
    x = temp;
  }
  return x || 1;
}

export default CoherenceDeltaScaffold;

